### PR TITLE
Collaborative Learning Algorithm added

### PR DESCRIPTION
**Pull request by sheelfshah**
There are a few `make lint` errors that I couldn't think a work around of.
MyCollaborativeMaster has been created to be able to create a list of masters with different parameters, and compare them in the same `game`

### CODE REVIEW

File: banditpylib/data.proto [hunk 1]
```diff
@@ -54,6 +54,18 @@ message Actions {
   repeated ArmPullsPair arm_pulls_pairs = 1;
 }
 
+// Next tag: 3
+message CollaborativeActions {
+  repeated ArmPullsPair arm_pulls_pairs = 1;
+
+  enum StateType {
```
Comment (Alanthink) [inline banditpylib/data.proto:61]: You can put `StateType` under `Actions` and add a comment `This is used by Collaborative learners`.

File: banditpylib/data.proto [hunk 2]
```diff
@@ -54,6 +54,18 @@ message Actions {
...
+    NORMAL = 0;   // wants to take actions [default]
```
Comment (Alanthink) [inline banditpylib/data.proto:62]: Please change `NORMAL` to `DEFAULT_NORMAL`

File: banditpylib/protocols/collaborative_learning_player.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+from banditpylib.data_pb2 import Trial, Actions
+from banditpylib.learners.collaborative_learner import CollaborativeBAILearner
+from .utils import Protocol
+
+
+class CollaborativeLearningProtocol(Protocol):
+  """Collaborative learning protocol
+
+  This protocol is used to simulate the multi-agent game
+  as discussed in the paper. It runs in rounds. During each round,
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_learning_player.py:17]: Paper citation

File: banditpylib/protocols/collaborative_learning_player.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  the protocol runs the following steps in sequence:
+
+  * fetch the state of the environment and ask each learner for actions;
+  * send the actions to the enviroment for execution;
+  * update each learner with the corresponding feedback of the environment;
+  * repeat the above steps until every agent enters the WAIT state;
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_learning_player.py:23]: or STOP state

File: banditpylib/protocols/collaborative_learning_player.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  * receive information broadcasted from every agent,
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_learning_player.py:24]: If there is at least one agent in WAIT state, then ...

File: changelog.txt [hunk 6]
```diff
@@ -0,0 +1,21 @@
+0. CollaborativeActions in data.proto
```
Comment (Alanthink) [inline changelog.txt:1]: Can you remove this file?

File: banditpylib/protocols/collaborative_protocol.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+from banditpylib.bandits import Bandit
+from banditpylib.data_pb2 import Trial, Actions
+from banditpylib.learners import Learner
+from banditpylib.learners.collaborative_learner import CollaborativeBAILearner
+from .utils import Protocol
+
+
+class CollaborativeLearningProtocol(Protocol):
+  """Collaborative learning protocol :cite:'tao2019collaborative'
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_protocol.py:15]: :cite:`tao2019collaborative`

File: banditpylib/protocols/collaborative_protocol.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+      bandits.append(dcopy(self.bandit))
+      bandits[-1].reset()
+
+    trial = Trial()
+    trial.bandit = self.bandit.name
+    trial.learner = current_learner.name
+
+    communication_rounds, total_pulls = 0, 0
+    active_agent_ids = list(range(len(agents)))
+    agent_arm_assignment = master.initial_arm_assignment(active_agent_ids)
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_protocol.py:73]: initial_arm_assignment accept number of agents.

File: banditpylib/learners/collaborative_learner/utils.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+  :param Optional[str] name: alias name
+  """
+
+  def __init__(self, name: Optional[str]):
+    self.__name = self._name() if name is None else name
+
+  @property
+  def name(self) -> str:
+    """Name of the agent"""
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:99]: Name of the master

File: banditpylib/learners/collaborative_learner/utils.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    return self.__name
+
+  @abstractmethod
+  def _name(self) -> str:
+    """
+    Returns:
+      default learner name
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:106]: Default master name

File: banditpylib/learners/collaborative_learner/utils.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:29]: default agent name

File: banditpylib/learners/collaborative_learner/utils.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    .. warning::
+      This function should be called before the start of each game.
+    """
+
+  @abstractmethod
+  def set_input_arms(self, arms: List[int]):
+    """Assign a set of arms to the agent
+
+    Args:
+      * arms: arm indices that have been assigneds
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:45]: There is no need to add * prefix.

File: banditpylib/learners/collaborative_learner/utils.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        :func:`actions` is executed
+    """
+
+  @property
+  @abstractmethod
+  def best_arm(self) -> int:
+    """Arm that the agent chose (could be None)"""
+
+  @property
+  def stage(self) -> str:
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:74]: Why you want this method?

Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/utils.py:74]: I think this is specific to the concrete agents.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+    horizon: int, num_agents: int, name: Optional[str] = None):
+    super().__init__(name)
+    if arm_num <= 1:
+      raise ValueError('Number of arms is expected at least 2. Got %d.' %
+                       arm_num)
+    if rounds <= 1:
+      raise ValueError('Number of rounds is expected at least 2. Got %d.' %
+                       rounds)
+    self.__arm_num = arm_num
+    self.__R = rounds - 1
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:312]: `self.__R` -> `self.__comm_rounds`

File: banditpylib/protocols/collaborative_protocol.py (hunk 8 above)
Comment (sheelfshah) [reply banditpylib/protocols/collaborative_protocol.py:73]: master already has num_agents, so I will pass no arguments

File: banditpylib/protocols/collaborative_protocol.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+  * if there is at least one agent in WAIT state, then receive information
+    broadcasted from every waiting agent, and use it to decide
+    the parameters of the next round.
+
+  The simulation stopping criteria is:
+
+  * every agent enters STOP state;
+
+  Algorithm is guaranteed to stop before total number of time-steps
+  achieve `horizon`.
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_protocol.py:34]: This sentence is not needed.

File: banditpylib/protocols/collaborative_protocol.py [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+  as discussed in the paper. It runs in rounds. During each round,
+  the protocol runs the following steps in sequence:
+
+  * fetch the state of the environment and ask each learner for actions;
+  * send the actions to the enviroment for execution;
+  * update each learner with the corresponding feedback of the environment;
+  * repeat the above steps until every agent enters the WAIT or STOP state;
+  * if there is at least one agent in WAIT state, then receive information
+    broadcasted from every waiting agent, and use it to decide
+    the parameters of the next round.
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_protocol.py:27]: ... then receive information broadcasted from every waiting agent and send them to master to decide arm assignment of next round.

File: banditpylib/protocols/collaborative_protocol.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+          arm_info = message_from_agent[arm_id]
+          new_pulls = messages[arm_id][1] + arm_info[1]
+          new_em_mean_reward = (messages[arm_id][1] * messages[arm_id][0] + \
+              arm_info[1] * arm_info[0]) / new_pulls
+          messages[arm_id] = (new_em_mean_reward, new_pulls)
+
+      # Send info to master for elimination to get arm assignment for next round
+      # agent_arm_assignment: key is agent_id, value is a list storing arm ids
+      # assigned to this agent
+      agent_arm_assignment = master.elimination(agent_in_wait_ids, messages)
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_protocol.py:126]: Can you put the logic of line 107 to line 121 inside the `master.elimination`? `master.elimination` will accept a dictionary whose key is agent id and value is a list of tuples which denote the information of the arms the agent pulls.

File: banditpylib/learners/utils.py [hunk 18]
```diff
@@ -1,7 +1,8 @@
 from abc import ABC, abstractmethod
 from typing import Optional, List, Union
 
-from banditpylib.data_pb2 import Context, Arm, Actions, Feedback
+from banditpylib.data_pb2 import (
+  Context, Arm, Actions, Feedback)
```
Comment (Alanthink) [inline banditpylib/learners/utils.py:5]: Why this change is needed?

File: banditpylib/learners/collaborative_learner/utils.py [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+    """Arm that the agent chose (could be None)"""
+
+  @abstractmethod
+  def broadcast(self) -> Dict[int, Tuple[float, int]]:
+    """Broadcasts information learnt in the current round
+
+    Returns: (a dict of)
+      * arm ids used in learning
+      * Tuple[corresponding average reward seen,
+        number of pulls used to deduce average]
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:80]: No need to use * prefix.

File: banditpylib/learners/collaborative_learner/utils.py [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+    return self.__agents
+
+  def get_master(self) -> CollaborativeBAIMaster:
+    """Controlling Master"""
+    return self.__master
+
+  @property
+  def goal(self) -> Goal:
+    arm = Arm()
+    if len(self.__master.active_arms) == 1:
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:181]: Ask the learner for the best arm not the master.

File: banditpylib/learners/collaborative_learner/utils.py [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+      messages: aggregation of messages broadcasted from agents
+
+    Returns:
+      dictionary of arm assignment per agent
+    """
+
+  @property
+  @abstractmethod
+  def active_arms(self):
+    """Arm indices that haven't been eliminated"""
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/utils.py:138]: This method is not needed?

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+    horizon: int, name: Optional[str] = None):
+    super().__init__(name)
+    if arm_num <= 1:
+      raise ValueError('Number of arms is expected at least 2. Got %d.' %
+                       arm_num)
+    if rounds <= 1:
+      raise ValueError('Number of rounds is expected at least 2. Got %d.' %
+                       rounds)
+    self.__arm_num = arm_num
+    self.__rounds = rounds - 1
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:156]: self.__comm_rounds

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+    self.__num_agents = num_agents
+
+  def _name(self) -> str:
+    return "lilUCBHeur_collaborative_master"
+
+  def reset(self):
+    self.__active_arms = list(range(self.__arm_num))
+
+  def __assign_arms(self, agent_ids: List[int]) ->\
+    Dict[int, List[int]]:
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:319]: Why do you want to break line?

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+      if np.random.uniform() > frac_x:
+        return int(x)
+      return int(x) + 1
+
+    if len(self.__active_arms) <= 1:
+      # don't assign any arms if 0/1 active arms
+      agent_arm_assignment = {}
+      for agent_id in agent_ids:
+        agent_arm_assignment[agent_id] = []
+      return agent_arm_assignment
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:336]: When there is only one arm, we should still assign arms such that each agent has the same arm.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 25]
```diff
@@ -0,0 +1,390 @@
...
+      arms_assign_list = []
+      num_agents_per_arm = len(agent_ids) / len(self.__active_arms)
+      for arm in self.__active_arms[:-1]:
+        arms_assign_list += [[arm]] * \
+          random_round(num_agents_per_arm)
+      arms_assign_list += [[self.__active_arms[-1]]] * \
+        (len(agent_ids) - len(arms_assign_list))
+      random.shuffle(arms_assign_list)
+      for i, agent_id in enumerate(agent_ids):
+        agent_arm_assignment[agent_id] = arms_assign_list[i]
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:348]: You could use `numpy.random.multinomial` to assign arms.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 26]
```diff
@@ -0,0 +1,390 @@
...
+      for arm in active_arms_copy[len(agent_ids):]:
+        agent_idx = random.choice(agent_ids)
+        agent_arm_assignment[agent_idx].append(arm)
+    return agent_arm_assignment
+
+  def initial_arm_assignment(self) -> Dict[int, List[int]]:
+    return self.__assign_arms(list(range(self.__num_agents)))
+
+  def elimination(self, agent_ids: List[int],
+    messages: Dict[int, Tuple[float, int]]) ->Dict[int, List[int]]:
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:366]: space after ->

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 27]
```diff
@@ -0,0 +1,390 @@
...
+    if not messages:
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:367]: When there is no message, we should raise exception.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py [hunk 28]
```diff
@@ -0,0 +1,390 @@
...
+      # terminate if no arms assigned
+      self.__stage = "termination"
+      return
+
+    self.__assigned_arms = np.array(arms)
+    # confidence of 0.01 suggested in the paper
+    self.__central_algo = LilUCBHeuristicCollaborative(self.__arm_num,
+      0.99, self.__assigned_arms)
+    self.__central_algo.reset()
+    self.__stage = "preparation"
```
Comment (Alanthink) [inline banditpylib/learners/collaborative_learner/collaborative_learning.py:188]: Why in set_input_arms, we always set the stage to `preparation`?

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 27 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:367]: It is possible that in learning, LilUCBHeuristic did not finish in horizon/2 time, which means agent return None in broadcast. So if all agents return None, then that means the algorithm failed (regret 1). Should we raise an Exception instead? What will it say - The horizon is insufficient to find the best arm ?

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 28 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:188]: setting input arms only occurs before every preparation step

Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:188]: I have changed it to 
```
if self.__stage == "unassigned":
      self.__stage = "preparation"
```

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 24 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:336]: Then how will the agents know that the algorithm has terminated and they should go to "termination" stage too

File: banditpylib/learners/collaborative_learner/utils.py (hunk 20 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/utils.py:181]: The algorithm says that the best arm is the only active arm or None

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 25 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:348]: I thought of using something similar too, but this might lead to randomness.
Eg. we might want to assign 4 arms to each agent, but end up assigning 2 to one and 6 to another.

Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:348]: Is it okay to if the number of arms assigned is random, but with the expectation same as `num_arms/num_agents`

File: banditpylib/learners/collaborative_learner/utils.py (hunk 21 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/utils.py:138]: Only needed if we get best arm from master

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 24 above)
Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:336]: They should monitor the number of rounds used and terminate when it reaches the last round.

File: banditpylib/protocols/collaborative_protocol.py [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+            break
+        max_pulls = max(max_pulls, pulls)
+      total_pulls += max_pulls
+
+      # stop if all agents are in STOP <=> no agents in WAIT
+      if not agent_in_wait_ids:
+        break
+
+      # communication and aggregation
+      # key is arm id and target is a dict broadcasted by agent
```
Comment (Alanthink) [inline banditpylib/protocols/collaborative_protocol.py:103]: key is agent id

File: banditpylib/learners/collaborative_learner/utils.py (hunk 20 above)
Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/utils.py:181]: Yes, I think it is talking about each agent. When the game stops, each agent is in state STOP. We should ask each agent her answer. If they are the same, the answer of the learner should be also the same. Otherwise, the learner has no answer.

File: banditpylib/learners/collaborative_learner/utils.py (hunk 21 above)
Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/utils.py:138]: See previous comment that we should ask agents for the final answer.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 25 above)
Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:352]: Okay, I think your method seems more robust. However, I think either way is fine. You can remain to use this method.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 27 above)
Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py]: You could randomly output one arm.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 24 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:336]: That is already being done, empty arm assignment is used for early stopping. So if we have only one active arm, then we can terminate the algorithm without doing all the learning steps

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 27 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py]: Are you saying LilUCBHeuristic should output a random arm
Or should the elimination and arm assignment step use a random arm

Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py]: LilUCBHeuristic should output a random arm I think

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 24 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:336]: We could also pass extra variables to `Protocol` in the `elimination` step, and these will be used in `set_input_arms` to decide early stopping

Comment (Alanthink) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:336]: Okay, you can remain the current implementation. Also please update the colab whenever the algorithm is changed.

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 27 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py]: We could initialize `LilUCBHeursiticCollaborative` with `horizon`, and it will return the currently best arm if `horizon` is reached

File: banditpylib/learners/collaborative_learner/collaborative_learning.py (hunk 24 above)
Comment (sheelfshah) [reply banditpylib/learners/collaborative_learner/collaborative_learning.py:336]: Since we are asking the agent for best_arm, this won't work.
I will try using some other method and push the changes

**Review by Alanthink — APPROVED** [review]

**Comment by Alanthink** [comment]
Collaborative learning algorithm added.
